### PR TITLE
Signup service

### PR DESCRIPTION
**Pull request by crufter**

### CODE REVIEW

**Comment by asim** [comment]
"Signup" service?

File: onboarding/handler/onboarding.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+	_, err = e.paymentService.CreateCustomer(ctx, &paymentsproto.CreateCustomerRequest{
+		Customer: &paymentsproto.Customer{
+			Id:   req.Email,
+			Type: "user",
+		},
+	})
+	return err
+}
+
+func (e *Onboarding) FinishOnboarding(ctx context.Context,
```
Comment (ben-toogood) [inline onboarding/handler/onboarding.go:186]: I'd rename this to something like StartSubscription

File: onboarding/proto/onboarding/onboarding.proto [hunk 2]
```diff
@@ -0,0 +1,45 @@
+syntax = "proto3";
+
+package go.micro.service.onboarding;
+
+service Onboarding {
+   rpc SendVerificationEmail(SendVerificationEmailRequest) returns (SendVerificationEmailResponse);
+   rpc Verify(VerifyRequest) returns (VerifyResponse);
+   // Creates a subscription and an account
+   rpc FinishOnboarding(FinishOnboardingRequest) returns (FinishOnboardingResponse);
```
Comment (asim) [inline onboarding/proto/onboarding/onboarding.proto:9]: Complete

File: onboarding/handler/onboarding.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+		return err
+	}
+	if tok.Token != req.Token {
+		return errors.New("Invalid token")
+	}
+
+	// If the user is verified, we are going to log her in and the
+	// flow stops here. We return the auth token to be used for further calls.
+	if tok.IsVerified {
+		secret, err := e.getAccountSecret(req.Email)
```
[the remark this answers is not in the document]
Comment (crufter) [reply onboarding/handler/onboarding.go]: Nice catch

File: onboarding/handler/onboarding.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+		}
+		token, err := e.auth.Token(auth.WithCredentials(req.Email, secret))
+		if err != nil {
+			return err
+		}
+		// @todo Is this correct?
+		rsp.AuthToken = token.RefreshToken
+		return nil
+	}
+	_, err = e.paymentService.CreateCustomer(ctx, &paymentsproto.CreateCustomerRequest{
```
[the remark this answers is not in the document]
Comment (crufter) [reply onboarding/handler/onboarding.go]: Thanks, done

File: signup/handler/signup.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+// Lifted  from the invite service https://github.com/micro/services/blob/master/projects/invite/handler/invite.go#L187
+// sendEmailInvite sends an email invite via the sendgrid API using the
+// predesigned email template. Docs: https://bit.ly/2VYPQD1
+func (e *Signup) sendEmail(email, token string) error {
+	logger.Infof("Sending email to address '%v'", email)
+
+	reqBody, _ := json.Marshal(map[string]interface{}{
+		"template_id": e.sendgridTemplateID,
+		"from": map[string]string{
+			"email": "Janos Dobronszki <dobronszki@gmail.com>",
```
Comment (jakexks) [inline signup/handler/signup.go:109]: Probably change this

Comment (crufter) [reply signup/handler/signup.go:109]: sure, haha

File: signup/handler/signup.go [hunk 6]
```diff
@@ -71,25 +91,34 @@ func (e *Signup) SendVerificationEmail(ctx context.Context,
 	logger.Info("Received Signup.SendVerificationEmail request")
 
 	// Save token
-	// @todo maybe use something nicer.
-	token := uuid.New().String()
+	k, err := totp.GenerateCodeCustom(e.key, time.Now(), totp.ValidateOpts{
```
Comment (jakexks) [inline signup/handler/signup.go:94]: totp token? Can people collide here by signing in at the same time?

Comment (crufter) [reply signup/handler/signup.go:94]: Very likely, I blindly followed Asim's code and I wish we would just use a random string vs something I'm unfamiliar with (otp entirely)

Comment (jakexks) [reply signup/handler/signup.go:94]: As discussed on slack, I would rather generate a short number derived from crypto/rand, keep it in the store with a TTL, then we can lose this dependency.

File: signup/handler/signup.go [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+		return err
+	}
+
+	return nil
+}
+
+// Lifted  from the invite service https://github.com/micro/services/blob/master/projects/invite/handler/invite.go#L187
+// sendEmailInvite sends an email invite via the sendgrid API using the
+// predesigned email template. Docs: https://bit.ly/2VYPQD1
+func (e *Signup) sendEmail(email, token string) error {
```
Comment (domwong) [inline signup/handler/signup.go:140]: At some point we need to introduce an email service and encapsulate sending emails in that but for MVP I'll allow it

File: signup/handler/signup.go [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+	req, err := http.NewRequest("POST", "https://api.sendgrid.com/v3/mail/send", bytes.NewBuffer(reqBody))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Authorization", "Bearer "+e.sendgridAPIKey)
+	req.Header.Set("Content-Type", "application/json")
+
+	if rsp, err := new(http.Client).Do(req); err != nil {
+		logger.Infof("Could not send email to %v, error: %v", email, err)
```
Comment (domwong) [inline signup/handler/signup.go:171]: return err

File: signup/handler/signup.go [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+	} else if rsp.StatusCode != 202 {
+		bytes, err := ioutil.ReadAll(rsp.Body)
+		if err != nil {
+			return err
+		}
+		logger.Infof("Could not send email to %v, error: %v", email, string(bytes))
```
Comment (domwong) [inline signup/handler/signup.go:177]: return err

File: signup/handler/signup.go (hunk 7 above)
Comment (crufter) [reply signup/handler/signup.go:140]: Yeah agreed

**Pull request merged**